File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols (VB)/FNet/ConfigurationCell.vb
'*******************************************************************************************************
'  ConfigurationCell.vb - FNet Cconfiguration cell
'  Copyright © 2008 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2008
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  02/08/2007 - J. Ritchie Carroll & Jian (Ryan) Zuo
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Text
Imports PhasorProtocols.Common
Imports PhasorProtocols.FNet.Common
Imports TVA.DateTime.Common

Namespace FNet

    <CLSCompliant(False), Serializable()> _
    Public Class ConfigurationCell

        Inherits ConfigurationCellBase

        Private m_ticksOffset As Long
        Private m_longitude As Single
        Private m_latitude As Single
        Private m_numberOfSatellites As Integer = 1 ' We'll initially assume synchronization is good until told otherwise

        Protected Sub New()
        End Sub

        ''' <summary>
        ''' Deserialize the configuration cell. Retrieve the Longitude, Latitude and the number of satellite
        ''' </summary>
        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

            ' Deserialize configuration cell
            m_ticksOffset = info.GetInt64("ticksOffset")
            m_longitude = info.GetSingle("longitude")
            m_latitude = info.GetSingle("latitude")
            m_numberOfSatellites = info.GetInt32("numberOfSatellites")

        End Sub

        Public Sub New(ByVal parent As ConfigurationFrame, ByVal nominalFrequency As LineFrequency, ByVal ticksOffset As Long)

            MyBase.New(parent, False, 0, nominalFrequency, MaximumPhasorValues, MaximumAnalogValues, MaximumDigitalValues)
            m_ticksOffset = ticksOffset

        End Sub

        Public Sub New(ByVal configurationCell As IConfigurationCell)

            MyBase.New(configurationCell)

        End Sub

        ' FNet supports no configuration frame in the data stream - so there will be nothing to parse

        '' This constructor satisfies ChannelCellBase class requirement:
        ''   Final dervived classes must expose Public Sub New(ByVal parent As IChannelFrame, ByVal state As IChannelFrameParsingState, ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32)
        'Public Sub New(ByVal parent As IConfigurationFrame, ByVal state As IConfigurationFrameParsingState, ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        '    '' We pass in defaults for id code and nominal frequency since these will be parsed out later
        '    'MyBase.New(parent, False, MaximumPhasorValues, MaximumAnalogValues, MaximumDigitalValues, _
        '    '    New ConfigurationCellParsingState( _
        '    '        AddressOf FNet.PhasorDefinition.CreateNewPhasorDefintion, _
        '    '        AddressOf FNet.FrequencyDefinition.CreateNewFrequencyDefintion, _
        '    '        Nothing, _
        '    '        Nothing), _
        '    '    binaryImage, startIndex)

        'End Sub

        'Friend Shared Function CreateNewConfigurationCell(ByVal parent As IChannelFrame, ByVal state As IChannelFrameParsingState(Of IConfigurationCell), ByVal index As Int32, ByVal binaryImage As Byte(), ByVal startIndex As Int32) As IConfigurationCell

        '    Return New ConfigurationCell(parent, state, index, binaryImage, startIndex)

        'End Function

        Public Overrides ReadOnly Property DerivedType() As System.Type
            Get
                Return Me.GetType
            End Get
        End Property

        Public Shadows ReadOnly Property Parent() As ConfigurationFrame
            Get
                Return MyBase.Parent
            End Get
        End Property

        ' FNet protocol only allows one device, so we share ID code with parent frame...
        Public Overrides Property IDCode() As UInt16
            Get
                Return Parent.IDCode
            End Get
            Set(ByVal value As UInt16)
                Parent.IDCode = value
            End Set
        End Property

        Public Property TicksOffset() As Long
            Get
                Return m_ticksOffset
            End Get
            Set(ByVal value As Long)
                m_ticksOffset = value
            End Set
        End Property

        Public Property Longitude() As Single
            Get
                Return m_longitude
            End Get
            Set(ByVal value As Single)
                m_longitude = value
            End Set
        End Property

        Public Property Latitude() As Single
            Get
                Return m_latitude
            End Get
            Set(ByVal value As Single)
                m_latitude = value
            End Set
        End Property

        Public Property NumberOfSatellites() As Integer
            Get
                Return m_numberOfSatellites
            End Get
            Set(ByVal value As Integer)
                m_numberOfSatellites = value
            End Set
        End Property

        Public Overrides ReadOnly Property MaximumStationNameLength() As Int32
            Get
                ' The station name is defined external to the protocol, so there is no set limit
                Return Int32.MaxValue
            End Get
        End Property

        ' FNet only supports floating point data
        Public Overrides Property PhasorDataFormat() As DataFormat
            Get
                Return DataFormat.FloatingPoint
            End Get
            Set(ByVal value As DataFormat)
                If value <> DataFormat.FloatingPoint Then Throw New NotSupportedException("FNet only supports floating point data")
            End Set
        End Property

        Public Overrides Property PhasorCoordinateFormat() As CoordinateFormat
            Get
                Return CoordinateFormat.Polar
            End Get
            Set(ByVal value As CoordinateFormat)
                If value <> CoordinateFormat.Polar Then Throw New NotSupportedException("FNet only supports polar coordinates")
            End Set
        End Property

        Public Overrides Property FrequencyDataFormat() As DataFormat
            Get
                Return DataFormat.FloatingPoint
            End Get
            Set(ByVal value As DataFormat)
                If value <> DataFormat.FloatingPoint Then Throw New NotSupportedException("FNet only supports floating point data")
            End Set
        End Property

        Public Overrides Property AnalogDataFormat() As DataFormat
            Get
                Return DataFormat.FloatingPoint
            End Get
            Set(ByVal value As DataFormat)
                If value <> DataFormat.FloatingPoint Then Throw New NotSupportedException("FNet only supports floating point data")
            End Set
        End Property

        ''' <summary>
        ''' Serialize the parameters of Longitude, Latitude and numberOfSatellite
        ''' </summary>
        Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

            MyBase.GetObjectData(info, context)

            ' Serialize configuration cell
            info.AddValue("ticksOffset", m_ticksOffset)
            info.AddValue("longitude", m_longitude)
            info.AddValue("latitude", m_latitude)
            info.AddValue("numberOfSatellites", m_numberOfSatellites)

        End Sub

        ''' <summary>
        ''' Retrieve the attribute include Longitude, Latitude and NumberOfSatellite
        ''' </summary>
        Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
            Get
                Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

                baseAttributes.Add("Time Offset (ticks)", m_ticksOffset)
                baseAttributes.Add("Time Offset (seconds)", m_ticksOffset / TicksPerSecond)
                baseAttributes.Add("Longitude", Longitude)
                baseAttributes.Add("Latitude", Latitude)
                baseAttributes.Add("Number of Satellites", NumberOfSatellites)

                Return baseAttributes
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/FNet/ConfigurationCell.cs
//*******************************************************************************************************
//  ConfigurationCell.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/08/2007 - James R Carroll & Jian (Ryan) Zuo
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Security.Permissions;

namespace PhasorProtocols.FNet
{
    /// <summary>
    /// Represents the F-NET implementation of a <see cref="IConfigurationCell"/> that can be sent or received.
    /// </summary>
    [Serializable()]
    public class ConfigurationCell : ConfigurationCellBase
    {
        #region [ Members ]

        // Fields
        private Ticks m_timeOffset;
        private double m_longitude;
        private double m_latitude;
        private int m_numberOfSatellites = 1; // We'll initially assume synchronization is good until told otherwise

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="ConfigurationCell"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The reference to parent <see cref="ConfigurationFrame"/> of this <see cref="ConfigurationCell"/>.</param>
        /// <param name="nominalFrequency">The nominal <see cref="LineFrequency"/> of the <see cref="FrequencyDefinition"/> of this <see cref="ConfigurationCell"/>.</param>
        /// <param name="timeOffset">The time offset of F-NET device in <see cref="Ticks"/>.</param>
        internal ConfigurationCell(ConfigurationFrame parent, LineFrequency nominalFrequency, Ticks timeOffset)
            : base(parent, parent.IDCode, Common.MaximumPhasorValues, Common.MaximumAnalogValues, Common.MaximumDigitalValues)
        {
            NominalFrequency = nominalFrequency;
            m_timeOffset = timeOffset;
        }

        /// <summary>
        /// Creates a new <see cref="ConfigurationCell"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected ConfigurationCell(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize configuration cell
            m_timeOffset = info.GetInt64("timeOffset");
            m_longitude = info.GetDouble("longitude");
            m_latitude = info.GetDouble("latitude");
            m_numberOfSatellites = info.GetInt32("numberOfSatellites");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets a reference to the parent <see cref="ConfigurationFrame"/> for this <see cref="ConfigurationCell"/>.
        /// </summary>
        public new ConfigurationFrame Parent
        {
            get
            {
                return base.Parent as ConfigurationFrame;
            }
            set
            {
                base.Parent = value;
            }
        }

        /// <summary>
        /// Gets or sets the ID code of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override ushort IDCode
        {
            // F-NET protocol only allows one device, so we share ID code with parent frame...
            get
            {
                return Parent.IDCode;
            }
            set
            {
                Parent.IDCode = value;
                base.IDCode = value;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> for the <see cref="IPhasorDefinition"/> objects in the <see cref="ConfigurationCellBase.PhasorDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <remarks>
        /// This property only supports floating-point data; F-NET doesn't transport scaled values.
        /// </remarks>
        /// <exception cref="NotSupportedException">F-NET only supports floating-point data.</exception>
        public override DataFormat PhasorDataFormat
        {
            get
            {
                return DataFormat.FloatingPoint;
            }
            set
            {
                if (value != DataFormat.FloatingPoint)
                    throw new NotSupportedException("F-NET only supports floating-point data");
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="CoordinateFormat"/> for the <see cref="IPhasorDefinition"/> objects in the <see cref="ConfigurationCellBase.PhasorDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <remarks>
        /// This property only supports polar phasor data; F-NET doesn't transport rectangular phasor values.
        /// </remarks>
        /// <exception cref="NotSupportedException">F-NET only supports polar phasor data.</exception>
        public override CoordinateFormat PhasorCoordinateFormat
        {
            get
            {
                return CoordinateFormat.Polar;
            }
            set
            {
                if (value != CoordinateFormat.Polar)
                    throw new NotSupportedException("F-NET only supports polar phasor data");
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> of the <see cref="FrequencyDefinition"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <remarks>
        /// This property only supports floating-point data; F-NET doesn't transport scaled values.
        /// </remarks>
        /// <exception cref="NotSupportedException">F-NET only supports floating-point data.</exception>
        public override DataFormat FrequencyDataFormat
        {
            get
            {
                return DataFormat.FloatingPoint;
            }
            set
            {
                if (value != DataFormat.FloatingPoint)
                    throw new NotSupportedException("F-NET only supports floating-point data");
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="DataFormat"/> for the <see cref="IAnalogDefinition"/> objects in the <see cref="ConfigurationCellBase.AnalogDefinitions"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        /// <remarks>
        /// This property only supports floating-point data; F-NET doesn't transport scaled values.
        /// </remarks>
        /// <exception cref="NotSupportedException">F-NET only supports floating-point data.</exception>
        public override DataFormat AnalogDataFormat
        {
            get
            {
                return DataFormat.FloatingPoint;
            }
            set
            {
                if (value != DataFormat.FloatingPoint)
                    throw new NotSupportedException("F-NET only supports floating-point data");
            }
        }

        /// <summary>
        /// Gets or sets time offset of F-NET device in <see cref="Ticks"/>.
        /// </summary>
        /// <remarks>
        /// F-NET devices normally report time in 11 seconds past real-time, this property defines the offset for this this artificial delay.
        /// Note that the parameter value is in ticks to allow a very high-resolution offset;  1 second = 10,000,000 ticks.
        /// </remarks>
        public Ticks TimeOffset
        {
            get
            {
                return m_timeOffset;
            }
            set
            {
                m_timeOffset = value;
            }
        }

        /// <summary>
        /// Gets or sets the longitude (i.e., East/West geospatial position) of the device defined by this <see cref="ConfigurationCell"/>.
        /// </summary>
        public double Longitude
        {
            get
            {
                return m_longitude;
            }
            set
            {
                m_longitude = value;
            }
        }

        /// <summary>
        /// Gets or sets the latitude (i.e., North/South geospatial position) of the device defined by this <see cref="ConfigurationCell"/>.
        /// </summary>
        public double Latitude
        {
            get
            {
                return m_latitude;
            }
            set
            {
                m_latitude = value;
            }
        }

        /// <summary>
        /// Gets or sets the number of satellites visible to the GPS timing source of the device defined by this <see cref="ConfigurationCell"/>.
        /// </summary>
        public int NumberOfSatellites
        {
            get
            {
                return m_numberOfSatellites;
            }
            set
            {
                m_numberOfSatellites = value;
            }
        }

        /// <summary>
        /// Gets the maximum length of the <see cref="ConfigurationCellBase.StationName"/> of this <see cref="ConfigurationCell"/>.
        /// </summary>
        public override int MaximumStationNameLength
        {
            get
            {
                // The station name is defined external to the protocol, so there is no set limit
                return int.MaxValue;
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="ConfigurationCell"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Time Offset (ticks)", ((long)TimeOffset).ToString());
                baseAttributes.Add("Time Offset (seconds)", Ticks.ToSeconds(TimeOffset).ToString());
                baseAttributes.Add("Longitude", Longitude.ToString());
                baseAttributes.Add("Latitude", Latitude.ToString());
                baseAttributes.Add("Number of Satellites", NumberOfSatellites.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize configuration cell
            info.AddValue("timeOffset", (long)m_timeOffset);
            info.AddValue("longitude", m_longitude);
            info.AddValue("latitude", m_latitude);
            info.AddValue("numberOfSatellites", m_numberOfSatellites);
        }

        #endregion
    }
}